### ISSUE TITLE
[Feature Request] Let's boost users' contributions in extending lammps

### ISSUE THREAD

**Issue by Amingeo** [lammps/lammps#1944]
**Summary**

Dear LAMMPS developers,
 Without any doubt, a remarkable percentage of LAMMPS users who have enjoyed working with its highly versatile and broad capabilities and have some programing experiences in c++, would be so ambitious to contribute in extending and developing LAMMPS, either to add new features related to their specific projects which might be useful for others or improving some previous algorithms and  features. Nevertheless, considering the high level of the programing applied in developing LAMMPS by developers who are typically experienced programmers and the versatile structure of the code, it might be a big challenge for new arrivals and specially not highly experienced c++ programmers to getting started and having themselves involved, resulting in being disappointed and surrender right at the beginning.
I am writing to ask/share with you a suggestion which can be very much helpful for such persons and that is, for few number of highly important .cpp files of LAMMPS, provide line by line elaborately explained comments and share it as some c++ or LAMMPS code tutorials. I would suppose, for majority of LAMMPS developers who know the structure of the code by heart, it won't take more than few minutes but it can save days/weeks of novices and will definitely have a very remarkable influences. 
I already surveyed a lot the web for something similar but couldn't find anything useful.
Bests,
Amin

**Comment by akohlmey** [comment]
Sorry, but your assessment of the situation is wrong in multiple ways.
First off, learning and research **is** painful, frustrating, and time consuming. The more complex something is, the more painful and the more time consuming and the more frustrating. You are spending so much time because you don't understand things. That is how it has always been. The biggest mistake that people can do is to aim too high too quickly. You need to build your expertise, experience and obtain (deserved) confidence. That will not happen quickly and most certainly not in the way you propose. Think of it like a math problem. You won't learn much if you get all the solutions and detailed explanations. You learn the most, when you have to figure things out step by step on your own, when you have to deal with problems that challenging (but not too challenging) so you can build your experience and expertise. Sooner or later you will discover that something that was difficult a while ago becomes simple as soon as you figured it out properly.

Second, you are misguided about how much effort it will take to do a detailed and line by line explanation. This is indeed **extremely** time consuming to do it properly and in a way that inexperienced people can follow.

Third, you are mixing C++ and general programming skills and understanding how things are structured in LAMMPS. Many of the core pieces in LAMMPS are reasonably well commented and - more importantly - are written in such a style, that it is actually straightforward to understand what it does.  If your C++ skills are not sufficient to understand the code, you need to lean more about C++ programming, and for that you don't need help from the LAMMPS developers. Not to mention, that we are far too few and have far too little time to give people C++ lectures. The programming style used in LAMMPS is in many places rather simple and - in my personal opinion - easy to follow and in a rather old fashioned, simple style of coding. The prototypical example is the lj/cut pair style. You can see what it does with just basic knowledge of C++ **and** a good enough understanding of the physics of MD and the Lennard-Jones potential in particular.  Both of which can be obtained from available text books.

Beyond that, a high-level view of the code structure and flow of control is given in the Developer.pdf file and some more details about how to and where to begin modifying LAMMPS in the corresponding sections of the LAMMPS manual. If you feel that that is not sufficient information, that it is most likely that you are missing some of the prerequisite skills I mentioned before and that is something that you can address yourself.

If you have trouble understanding parts of LAMMPS, you also have the opportunity to ask (well formed) questions on the lammps-users mailing list. Lots of people have done this, sometimes it took a while, sometimes they could obtain the information they were looking for quickly.

If it would be so difficult to write code to be included in LAMMPS, we would have so many contributions. In fact, compared to many other codes in the field (say, Gromacs or NAMD) it is conceivably much simpler to add features to LAMMPS or modify it.  The vast majority of code in LAMMPS is *not* written by the LAMMPS (core) developers but by contributors (although occasionally a contributor may become a (core) developer).

Please let me add, that I regularly teach classes in programming algorithms and data structures to scientists and am guiding people how to develop scientific software for over 15 years. Thus I feel that by now have a good understanding of how people are best guided towards becoming better programmers and becoming able to implement complex tasks.  What you propose is not very effective. The most important skill a programming needs is to read and understand some other programmer's code and the ability to write code, so that it expresses what it does (the comments should only explain *why* certain choices are made, not *how* things are done).

If you worry about saving days or weeks, you are not understanding what you are up against. Good understanding does not come over night. it requires perseverance and the ability to gradually build a skill and not demand that the answer will be handed to you with all detailed explanations.  Figuring something out is actually the fun part of programming (and research). Those are the 10% of the time you enjoy within the 90% of frustration because you are stuck with something that you don't understand yet. The experienced programmer will know and be confident, that he sooner or later will figure it out. Sometimes it requires a different perspective or discussing your insights with somebody. I've personally have made the experience that that somebody doesn't even need to know much about the problem you are stuck with. Often the simple procedure of explaining a problem to somebody else can put you in the right path to solve it and get out of corners you are stuck in.

Now this turned into a rather long essay, but it is my personal view of things based on my personal experience as a teacher **and** as a student. I have been at times as much frustrated as you seem to be, but I have also over time learned to be patient and try to break down big insurmountable problems into smaller, easier to solve pieces and thus - gradually - understand the big picture and thus solve the actual problem at hand. 

I have heard similar suggestions (not just related to LAMMPS) repeatedly and usually the argument is made, that there could be more contributions.  However, that is missing the point. There is no intrinsic benefit in making LAMMPS the project with the most lines of codes, or the most features, or the one MD code to end all MD codes, if those lines of code are written by inexperienced and incompletely trained people and the features don't work so well because of that and the maintenance of the code become even more of a burden as it already is.

Hope that explains some things and will encourage you to not give up too easily. Giving up easily won't help with programming and won't help with research in general. And let me repeat, the frustration you feel is part of the process. If you feel lost, set your aim lower and try something simpler. And make sure you have acquired the sufficient prerequisite skills.

Or to just put it into a simple phrase:  There ain't no escape from the blues.  ;-)

**Comment by Amingeo** [comment]
Thanks Axel for your comment. I totally agree with all your statements. However, I think you misunderstood what I was suggesting. What I was suggesting was sharing well explained examples for "few" cpp codes (1-5 examples). Obviously, people who have shallow c++ experience cannot and are not going to learn it this way (and also the LAMMPS source code). It anyways needs all hard efforts that you talked about and I also agreed. It will be just a clue and fagend to help interested persons to get started, otherwise, those developer manuals also wouldn't have been published ...

**Comment by akohlmey** [comment]
Since those 1-5 example files would need to be different for every person, who is going to pick which? Please provide some examples of files or parts of files that a) you are struggling to understand in order to write new code to be added to LAMMPS and b) where the relevant information to understand what the code is doing cannot be obtained from the existing information I mentioned and through having a sufficient understanding of the necessary background in programming as well as statistical mechanics and physics.

And you significantly underestimate the effort required.  It is most easily done with new files and features, and for that I can name the files `utils.h` and `utils.cpp`, which I have "decorated" with additional comments to demonstrate, what it would look like, if we would ever start writing a "programmer's guide" by utilizing the doxygen software to extract and visualize the embedded information and semi-automatically also document the code structure.

In order to assess the overall effort of providing detailed , I have (long ago) done the same for a *complete* small project of mine called [TclMPI](https://github.com/akohlmey/tclmpi). It has indeed been quite a major undertaking despite the rather limited scope.  I would **love** to have something similar done for LAMMPS, but we are just a few people and have to pick our priorities. There are so many other things to do that have a much deeper impact to the LAMMPS community (users **and** developers).

**Comment by Amingeo** [comment]
> Please provide some examples of files or parts of files that a) you are struggling to understand in order to write new code to be added to LAMMPS
Please let me to declare that I am not suggesting it only on behalf of myself. I can apply new things I need to try by modifying some in house codes written in fortran90. For LAMMPS modifying, I already have done this once without needing to understand the whole code and only via changing some similar codes. The reason I suggested it is that I learned that this is exactly something many colleagues and students wish to and makes getting started ways easier. 
Although you said most of .cpp codes of lammps are well commented, I have not seen it and can say they are minimally commented (see e.g. fix_temp_rescale.cpp).

 

> It is most easily done with new files and features, and for that I can name the files utils.h and utils.cpp, which I have "decorated" with additional comments to demonstrate, what it would look like, if we would ever start writing a "programmer's guide" by utilizing the doxygen software to extract and visualize the embedded information and semi-automatically also document the code structure.

Sounds great. And to make it easier, even if it is not line by line extensively commented (which I guess I kind of overstated what I meant), even few comments for each part of the code to declare what it is supposed to and how it connects to other classes for few important parts (update.cpp, fix_npt.cpp, ..) would be a great help and probably more helpful to understand the structure of the code compared to developer manuals. That's what I think and of course only a suggestion.

**Comment by akohlmey** [comment]
> Please let me to declare that I am not suggesting it only on behalf of myself. I can apply new things I need to try by modifying some in house codes written in fortran90. For LAMMPS modifying, I already have done this once without needing to understand the whole code and only via changing some similar codes. The reason I suggested it is that I learned that this is exactly something many colleagues and students wish to and makes getting started ways easier. At least for most of .cpp codes of lammps I can say they are minimally commented (see e.g. fix_temp_rescale.cpp).

You are *still* missing my point. There is no way to bypass having to go through the process of learning the way how I have already outlined it. You can call it "the hard way", see the website and book "Learn Python the Hard Way" for and example and good arguments, why this is the better way to approach writing code). Fix `temp/rescale` is a good example for a file that has **exactly** the right amount of comments in the right places.  The remaining information is already in the manual and in the Developer.pdf document that describes the overall flow of control.  Asking for more is ignoring the realities of how much time LAMMPS (core) developers have to spend on working on LAMMPS and how few of them there are that can do this kind of work. While what you are proposing may get more people started, it will give them a wrong sense of their own skills and that is very dangerous. If you the search the web for the essay "How Developers Stop Learning: Rise of the Expert Beginner" and some more related texts on "expert beginners", you find more examples and better arguments than I have the time to present here, why it is sometimes in the grand scheme of things better to make things not too easy. I think this very much applies here.


> 
> Sounds great. And to make it easier, even if it is not line by line extensively commented (which I guess I kind of overstated what I meant), even few comments for each part of the code to declare what it is supposed to and how it connects to other classes for few important parts (update.cpp, fix_npt.cpp, ..) would be a great help and probably more helpful to understand the structure of the code compared to developer manuals. 

The overall flow of control and how everything is connected together *is* outlined in Developer.pdf. There is not much to explain about fix npt.cpp its relation to fix_nh.cpp requires just basic C++ knowledge and the integration of a fix into the flow of control is - again - outlined in Developer.pdf and the purpose of the individual required functions of a fix in the manual.  That aside, the nose-hoover fixes are a rather unusual part of LAMMPS and to understand the functioning of the time stepping, it is sufficient to study fix nve instead. The rest is specific to nose-hoover thermostat and barostat algorithms and thus very specific and therefore not something for a beginner to start with.

Don't hold you breath on more doxygen decorations in the manual. As I already outlined, this is a low priority item and will take a very long time to get anywhere to become useful. 

We are starting to argue in circles. I do understand the frustrations of some people (I have been there myself), but I have to re-state that this something that primarily needs to be addressed by a change of the approach and adjustment of expectations of what can be done with a given level of experience.  If anybody is in a position to assist with that, those would be the individual PIs and supervisors that need to assert suitable advice and guidance on their coworkers and students. We are seeing time and again, that those rather than providing that kind of guidance (like I have experienced it from my adviser and by whom I was given the time to learn things "the hard way(tm)"), and from collaborators on various software packages) revolve to abandoning their students/posdocs/coworkers for fear of exposing their own lack of experience and creating pressure to produce results.

To summarize. The LAMMPS developers are aware of how daunting it is to start working on as large a software package as LAMMPS.  We are committed to keep at least the current status (even though it is a challenge already) and have raised our requirements and automated tests for code quality (to existing and new code) over time already.  Time permitting, we will work on improving the situation. However, there is plenty of guidance already available and the vast majority of problems people are having originate from impatience or lack of experience, neither of which can be addressed by the LAMMPS developers through expanding documentation and comments in the source code.

**Comment by akohlmey** [comment]
New programmer guide section is added to the manual. will be expanded/improved over time.
